boot/bootutil/src/loader.c

```diff
@@ -1549,89 +1559,432 @@ boot_set_image_ok(void)
 #endif /* !MCUBOOT_OVERWRITE_ONLY */
 
 /**
- * Performs an image swap if one is required.
+ * Sets the active image on which the bootloader will continue to work.
+ *
+ * @param image_id                  The index of the image (from 0).
+ */
+static void
+boot_set_active_image(size_t image_id)
+{
+    current_image = image_id;
+    boot_data = &boot_data_array[image_id];
+
+#ifdef MCUBOOT_ENC_IMAGES
+    /* The keys used for encryption are no longer valid (may belong to
+     * another image).
+     */
+    boot_enc_zeroize();
```

Comment:
I *think* this would only be required a single time (per image); maybe it could be done once before any swap is operated? Also right now there would be a single key for any number of images, so the only thing that would require reset would possibly be the `valid` state.

Reply:
I've created a boot_enc_mark_keys_invalid() function instead of using the boot_enc_zeroize(). In case of multi image boot it must be called in each main for cycles (in boot_go) where it swaps images or trying to validate the image in the secondary slot.